src/pyodide/internal/python.js

```diff
@@ -278,14 +284,43 @@ function mountLib(pyodide) {
 }
 
 export async function loadPyodide() {
+  // Lookup memory snapshot from artifact store.
+  const maybeMemorySnapshot = ArtifactBundler.getMemorySnapshot();
+  if (maybeMemorySnapshot) {
+    // Simple sanity check to ensure this snapshot isn't corrupted.
+    //
+    // TODO(later): we need better detection when this is corrupted. Right now the isolate will
+    // just die.
+    if (maybeMemorySnapshot.length > 100) {
+      MEMORY = maybeMemorySnapshot;
+    }
+  }
+
   const emscriptenSettings = getEmscriptenSettings();
   const emscriptenModule = await instantiateEmscriptenModule(
     emscriptenSettings
   );
-  prepareWasmLinearMemory(emscriptenModule);
+  const usingExistingMemorySnapshot = await prepareWasmLinearMemory(emscriptenModule);
+
+  // Upload `MEMORY` to artifact store so long as a new memory snapshot was generated.
+  if (!ArtifactBundler.isDisabled() && !usingExistingMemorySnapshot) {
```

Comment:
How about removing the double negative?
```suggestion
  if (ArtifactBundler.isEnabled() && !usingExistingMemorySnapshot) {
```